### ISSUE TITLE
Problems with SessionHandlerDB and ajax images upload and "tempFiles"

### ISSUE THREAD

**Issue by somatonic** [ryancramerdesign/ProcessWire#1871]
I found out after debugging why on some installations the images ajax upload is broken regarding isTemp and its storage in session.

It turns out it happens only if the SessionHandlerDB is installed. 

So if I drag some images into the images field they get uploaded fine as usual, but after saving the page some images disappear. While they're still in the assets folder they are not saved to the field for some reasons.

I found that it happens the more and the bigger the images are. So if I upload 3-4 small jpg it may work, but as soon as I upload 3-4 big jpgs (1-2mb), half of them tend to get lost.

 When I tried to debug this one, I found that the files are added to the tempFiles session key, but not correctly it seems. I upload 3 images without saving the page, the tempFiles array only holds 1 image in the session, while it should be all 3 in there.

This happens on different installs, different servers and on my local server too. PW 2.7 and PW 3 as soon as SessioHandlerDB is installed.

**Comment by ryancramerdesign** [comment]
I use SessionHandlerDB on pretty much all of my installs, but have not observed this issue. Just made a couple attempts with seven 5-7mb images that took awhile to upload, but everything seems to be working how it should. Are there any other factors present that you can think of? Any other modules making ajax requests in the background, like SystemNotifications for instance? Are you editing other pages on the site while the files upload, or leaving it alone?

**Comment by somatonic** [comment]
No vanilla PW, no modules not other actions. And it seems others can reproduce this problem too.

**Comment by FlipZoomMedia** [comment]
I can confirm the problem. But I could not explain to me, how and why. I've blame it first to the PHP max_upload_size, but that was wrong. I have noticed this last week, with a new PW 3.x installation without modules, but also on an existing 2.7 installation with an upload field for a gallery. Uploads with pictures greater then 2 MB the half is lost.

**Comment by arjenblokzijl** [comment]
Same issues here too. Clean PW 3.0.21 install with Minimal Site Profile on MySQL 5.5.49 and PHP 7.0.

Steps I've taken:

1. Create an image field and assign to the homepage
2. Upload bunch of images (I used 72 images from around 1mb to 5mb)
3. When the uploading is finished save the page
4. Refresh page => all 72 images are there (so far so good 👍 )
5. Delete images
6. Install SessionHandlerDB
7. Re-run steps from 1-4 and suddenly there are 66 images.

**Comment by adrianbj** [comment]
Just referencing a related forum post: https://processwire.com/talk/topic/13522-problem-with-ajax-images-upload-and-sessionhandlerdb/

**Comment by LostKobrakai** [comment]
This seems like a race condition issue to me. I'm just not sure, why this should be correlated to SessionHandlerDB.

I've added this to the Pagefiles::isTemp method right after $tempFiles retrieval from the session.
`$this->log->save('debug', count($tempFiles));`

```
2016-06-12 21:08:14	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	0
2016-06-12 21:08:18	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	1
2016-06-12 21:08:24	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	2
2016-06-12 21:08:25	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	2
2016-06-12 21:08:26	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	2
2016-06-12 21:08:27	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	3
2016-06-12 21:08:30	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	4
2016-06-12 21:08:33	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	5
…
2016-06-12 21:09:17	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	22
2016-06-12 21:09:20	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	23
2016-06-12 21:09:22	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	23
2016-06-12 21:09:22	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	24
…
2016-06-12 21:09:31	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	27
2016-06-12 21:09:33	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	28
2016-06-12 21:09:36	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	28
2016-06-12 21:09:36	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	29
…
2016-06-12 21:09:44	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	32
2016-06-12 21:09:48	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	33
2016-06-12 21:09:49	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	33
2016-06-12 21:09:51	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	33
2016-06-12 21:09:52	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	34
…
2016-06-12 21:10:27	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	48
2016-06-12 21:10:30	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	49
2016-06-12 21:10:32	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	49
2016-06-12 21:10:32	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	50
…
2016-06-12 21:10:43	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	54
2016-06-12 21:10:45	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	55
2016-06-12 21:10:48	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	55
2016-06-12 21:10:48	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	56
…
2016-06-12 21:10:57	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	59
2016-06-12 21:11:00	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	60
2016-06-12 21:11:02	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	60
2016-06-12 21:11:03	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	61
…
2016-06-12 21:11:11	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	64
2016-06-12 21:11:14	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	65
2016-06-12 21:11:16	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	65
2016-06-12 21:11:16	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	66
…
2016-06-12 21:11:57	benni	http://images.valet/manage/page/edit/?id=1&InputfieldFileAjax=1	82
```

**Comment by ryancramerdesign** [comment]
I think I've tracked down the issue. I was able to duplicate it on localhost after attempting to upload 84 giant jpegs, because I had a few missing after that. The ajax uploads are just throwing too many requests to the server at the same time. Each request has to obtain a lock for the session data, otherwise it would become corrupted. Some of the uploads aren't able to obtain an exclusive lock before write, causing their info to go missing. 

The issue is that it had a 10-second retry period, which just wasn't long enough to obtain a lock when there are dozens of connections being made at once. I've bumped it up to a 50-second retry period. This solved it here, but it may or may not solve it there, because the fact remains that this is too many requests to throw at once. So a better long term solution will be for us to have the ajax file uploads throttle themselves, finishing one file before moving on to the next. That way it won't attempt to open 70+ (or how ever many) simultaneous connections to the server at the same time. Btw, the quantity doesn't really matter that much, what matters is how long it takes. 

I still couldn't duplicate it on an actual web server because I think the server I'm testing with must have apache limiting the number of simultaneous connections, and staying within a limit that's safe for uploads. Whereas MAMP (local web server) doesn't seem to be placing any limits.
